src/Microsoft.AspNetCore.Sockets/HttpConnectionDispatcher.cs

```diff
@@ -49,128 +45,141 @@ public HttpConnectionDispatcher(ConnectionManager manager, ILoggerFactory logger
             else if (context.Request.Path.StartsWithSegments(path + "/poll"))
             {
-                // Connection must already exist
-                var state = await GetConnectionAsync(context);
-                if (state == null)
-                {
-                    // No such connection, GetConnection already set the response status code
-                    return;
-                }
-
-                if (!await EnsureConnectionStateAsync(state, context, LongPollingTransport.Name))
-                {
-                    // Bad connection state. It's already set the response status code.
-                    return;
-                }
+                await ExecuteLongPollingAsync(context, endpoint, connectionState);
+            }
+            else if (context.Request.Path.StartsWithSegments(path + "/send"))
+            {
+                await ExecuteSendAsync(context, connectionState);
+            }
+        }
 
-                // Mark the connection as active
-                state.Active = true;
+        private async Task ExecuteTransportAsync(HttpContext context, EndPoint endpoint, ConnectionState connectionState, IHttpTransport transport)
+        {
+            // Set up connection state
+            if (!EnsureConnectionState(connectionState, context, transport.Name))
+            {
+                // Changed transports mid-connection. Bad client!
+                context.Response.StatusCode = 400;
+                await context.Response.WriteAsync("Cannot change transports mid-connection");
+                return;
+            }
 
-                var longPolling = new LongPollingTransport(state.Application.Input, _loggerFactory);
+            await DoPersistentConnection(endpoint, transport, context, connectionState);
+            _manager.RemoveConnection(connectionState.Connection.ConnectionId);
+        }
 
-                // Start the transport
-                var transportTask = longPolling.ProcessRequestAsync(context);
+        private async Task ExecuteLongPollingAsync(HttpContext context, EndPoint endpoint, ConnectionState connectionState)
+        {
+            // Mark the connection as active
+            connectionState.Active = true;
 
-                // Raise OnConnected for new connections only since polls happen all the time
-                var endpointTask = state.Connection.Metadata.Get<Task>("endpoint");
-                if (endpointTask == null)
-                {
-                    _logger.LogDebug("Establishing new Long Polling connection: {0}", state.Connection.ConnectionId);
+            var longPolling = new LongPollingTransport(connectionState.Application.Input, _loggerFactory);
 
-                    // This will re-initialize formatType metadata, but meh...
-                    state.Connection.Metadata["transport"] = LongPollingTransport.Name;
+            // Set up connection state
+            if (!EnsureConnectionState(connectionState, context, longPolling.Name))
```

Comment:
Except that it needs to be in each separate branch because we need to know what transport the connection is using before we can do it.